src/compiler/checker.ts

```diff
@@ -5450,11 +5524,32 @@ module ts {
 
                 return type;
             }
+            
+            function narrowTypeByCallExpression(type: Type, expr: CallExpression, assumeTrue: boolean): Type {
+                // Check that type is not any, assumed result is true.
+                if (type.flags & TypeFlags.Any || !assumeTrue) {

```

Comment:
If assumeTrue is false and type is a union, I think you can remove constituents that are subtypes of the guard. That is what we do for instanceOf guards.